cmd/account/mgmt/account-assign.go

```diff
@@ -124,12 +130,18 @@ func (o *accountAssignOptions) findUntaggedAccount(rootOu string) (string, error
 	if err != nil {
 		return "", err
 	}
+
 	if len(accounts.Accounts) == 0 {
-		return "", ErrNoAccountsInRoot
+		fmt.Println("No accounts available to assign")
+
+		accountAssignID, err := o.buildAccount(seedVal)
+		if err != nil {
+			return "", err
+		}
+		return accountAssignID, nil
```

Comment:
This in general is a code smell, but I'm not going to ask for this to be tackled here, this is a good candidate for a refactor.

The function that this body of work is included in is called `findUntaggedAccount`.  Creating a new account if there is no account available is a side-effect, and _can_ get messy when attempting to debug.

I think a better pattern would be to do something on line 98/99 like this:

```go
accountAssignID, err := o.findUntaggedAccount(rootID)
if err != nil {
  # If the error returned is not because of a lack of accounts, return the error
  if err != ErrNoUntaggedAccounts {
    return "", err
  }
  # otherwise, create a new account
  seed := time.Now().UnixNano()
  accountAssignID, err = o.buildAccount(seed)
  if err != nil {
    return "", err
  }
}
```

Again, I'm not going to ask for this change now, and looking at the code I just wrote I'm not sure which is easier to understand at first, but I think the main difference is that we're not relying on a side-effect functionality as part of our core logic.

Reply:
+100